File: web/src/services/api/subscription.ts
import { apiDelete, apiGet, apiPost, apiPut, compactApiParams } from "@/services/api/request";

export type SubscriptionDurationUnit = "day" | "month" | "year" | "hour" | "custom";
export type SubscriptionResetCycle = "none" | "daily" | "weekly" | "monthly" | "custom";

export type SubscriptionPlan = {
    id: string;
    title: string;
    subtitle: string;
    priceCredits: number;
    upgradeRole: string;
    downgradeRole: string;
    purchaseLimit: number;
    sort: number;
    enabled: boolean;
    durationUnit: SubscriptionDurationUnit;
    durationValue: number;
    durationCustomSeconds: number;
    quotaCredits: number;
    resetCycle: SubscriptionResetCycle;
    resetCustomSeconds: number;
    allowWalletFallback: boolean;
    subscriberCount: number;
    createdAt: string;
    updatedAt: string;
};

export type SubscriptionPlanList = {
    items: SubscriptionPlan[];
    total: number;
};

export type UserSubscription = {
    id: string;
    userId: string;
    planId: string;
    planTitle: string;
    priceCredits: number;
    upgradeRole: string;
    downgradeRole: string;
    quotaCredits: number;
    quotaRemaining: number;
    resetCycle: SubscriptionResetCycle;
    resetCustomSeconds: number;
    allowWalletFallback: boolean;
    source: "purchase" | "admin";
    status: "active" | "expired" | "replaced" | "voided";
    startsAt: string;
    expiresAt: string;
    lastResetAt: string;
    nextResetAt: string;
    createdAt: string;
    updatedAt: string;
};

export type UserSubscriptionList = {
    items: UserSubscription[];
    total: number;
};

export type SubscriptionPurchaseResult = {
    user: import("@/services/api/auth").AuthUser;
    subscription: UserSubscription;
};

export type SubscriptionSubscriber = {
    subscriptionId: string;
    userId: string;
    username: string;
    displayName: string;
    quotaCredits: number;
    quotaRemaining: number;
    status: UserSubscription["status"];
    startsAt: string;
    expiresAt: string;
};

export type SubscriptionSubscriberList = {
    items: SubscriptionSubscriber[];
    total: number;
};

export async function fetchAdminSubscriptionPlans(token: string, query: { keyword?: string; status?: string; page?: number; pageSize?: number } = {}) {
    return apiGet<SubscriptionPlanList>("/api/admin/subscription-plans", compactApiParams(query), token);
}

export async function createAdminSubscriptionPlan(token: string, data: Partial<SubscriptionPlan>) {
    return apiPost<SubscriptionPlan>("/api/admin/subscription-plans", data, token);
}

export async function updateAdminSubscriptionPlan(token: string, id: string, data: Partial<SubscriptionPlan>) {
    return apiPut<SubscriptionPlan>(`/api/admin/subscription-plans/${encodeURIComponent(id)}`, data, token);
}

export async function deleteAdminSubscriptionPlan(token: string, id: string) {
    return apiDelete<boolean>(`/api/admin/subscription-plans/${encodeURIComponent(id)}`, token);
}

export async function fetchAdminSubscriptionPlanUsers(token: string, planId: string, query: { keyword?: string; page?: number; pageSize?: number } = {}) {
    return apiGet<SubscriptionSubscriberList>(`/api/admin/subscription-plans/${encodeURIComponent(planId)}/users`, compactApiParams(query), token);
}

export async function fetchAdminUserSubscriptions(token: string, userId: string) {
    return apiGet<UserSubscriptionList>(`/api/admin/users/${encodeURIComponent(userId)}/subscriptions`, { page: 1, pageSize: 100 }, token);
}

export async function grantAdminUserSubscription(token: string, userId: string, planId: string) {
    return apiPost<UserSubscription>(`/api/admin/users/${encodeURIComponent(userId)}/subscriptions`, { planId }, token);
}

export async function resetAdminUserSubscription(token: string, subscriptionId: string) {
    return apiPost<UserSubscription>(`/api/admin/user-subscriptions/${encodeURIComponent(subscriptionId)}/reset`, {}, token);
}

export async function voidAdminUserSubscription(token: string, subscriptionId: string) {
    return apiPost<boolean>(`/api/admin/user-subscriptions/${encodeURIComponent(subscriptionId)}/void`, {}, token);
}

export async function deleteAdminUserSubscription(token: string, subscriptionId: string) {
    return apiDelete<boolean>(`/api/admin/user-subscriptions/${encodeURIComponent(subscriptionId)}`, token);
}

export async function fetchSubscriptionPlans(token: string) {
    return apiGet<SubscriptionPlanList>("/api/v1/subscription-plans", { page: 1, pageSize: 100 }, token);
}

export async function fetchUserSubscriptions(token: string) {
    return apiGet<UserSubscriptionList>("/api/v1/subscriptions", { page: 1, pageSize: 100 }, token);
}

export async function purchaseSubscription(token: string, planId: string) {
    return apiPost<SubscriptionPurchaseResult>("/api/v1/subscriptions/purchase", { planId }, token);
}

File: model/subscription.go
package model

type SubscriptionDurationUnit string

const (
	SubscriptionDurationDay    SubscriptionDurationUnit = "day"
	SubscriptionDurationMonth  SubscriptionDurationUnit = "month"
	SubscriptionDurationYear   SubscriptionDurationUnit = "year"
	SubscriptionDurationHour   SubscriptionDurationUnit = "hour"
	SubscriptionDurationCustom SubscriptionDurationUnit = "custom"
)

type SubscriptionResetCycle string

const (
	SubscriptionResetNone    SubscriptionResetCycle = "none"
	SubscriptionResetDaily   SubscriptionResetCycle = "daily"
	SubscriptionResetWeekly  SubscriptionResetCycle = "weekly"
	SubscriptionResetMonthly SubscriptionResetCycle = "monthly"
	SubscriptionResetCustom  SubscriptionResetCycle = "custom"
)

type SubscriptionPlan struct {
	ID                    string                   `json:"id" gorm:"primaryKey"`
	Title                 string                   `json:"title"`
	Subtitle              string                   `json:"subtitle"`
	PriceCredits          int                      `json:"priceCredits"`
	UpgradeRole           string                   `json:"upgradeRole" gorm:"index"`
	DowngradeRole         string                   `json:"downgradeRole"`
	PurchaseLimit         int                      `json:"purchaseLimit" gorm:"default:0"`
	Sort                  int                      `json:"sort" gorm:"default:0"`
	Enabled               bool                     `json:"enabled" gorm:"default:true;index"`
	DurationUnit          SubscriptionDurationUnit `json:"durationUnit"`
	DurationValue         int                      `json:"durationValue"`
	DurationCustomSeconds int                      `json:"durationCustomSeconds" gorm:"default:0"`
	QuotaCredits          int                      `json:"quotaCredits" gorm:"default:0"`
	ResetCycle            SubscriptionResetCycle   `json:"resetCycle" gorm:"default:none"`
	ResetCustomSeconds    int                      `json:"resetCustomSeconds" gorm:"default:0"`
	AllowWalletFallback   bool                     `json:"allowWalletFallback" gorm:"default:false"`
	CreatedAt             string                   `json:"createdAt"`
	UpdatedAt             string                   `json:"updatedAt"`
	SubscriberCount       int                      `json:"subscriberCount" gorm:"-"`
}

type SubscriptionPlanList struct {
	Items []SubscriptionPlan `json:"items"`
	Total int                `json:"total"`
}

type UserSubscriptionStatus string

const (
	UserSubscriptionActive   UserSubscriptionStatus = "active"
	UserSubscriptionExpired  UserSubscriptionStatus = "expired"
	UserSubscriptionReplaced UserSubscriptionStatus = "replaced"
	UserSubscriptionVoided   UserSubscriptionStatus = "voided"
)

type UserSubscriptionSource string

const (
	UserSubscriptionSourcePurchase UserSubscriptionSource = "purchase"
	UserSubscriptionSourceAdmin    UserSubscriptionSource = "admin"
)

type UserSubscription struct {
	ID                  string                 `json:"id" gorm:"primaryKey"`
	UserID              string                 `json:"userId" gorm:"index"`
	PlanID              string                 `json:"planId" gorm:"index"`
	PlanTitle           string                 `json:"planTitle"`
	PriceCredits        int                    `json:"priceCredits"`
	UpgradeRole         string                 `json:"upgradeRole"`
	DowngradeRole       string                 `json:"downgradeRole"`
	QuotaCredits        int                    `json:"quotaCredits"`
	QuotaRemaining      int                    `json:"quotaRemaining"`
	ResetCycle          SubscriptionResetCycle `json:"resetCycle"`
	ResetCustomSeconds  int                    `json:"resetCustomSeconds"`
	AllowWalletFallback bool                   `json:"allowWalletFallback"`
	Source              UserSubscriptionSource `json:"source" gorm:"default:purchase"`
	Status              UserSubscriptionStatus `json:"status" gorm:"index"`
	StartsAt            string                 `json:"startsAt"`
	ExpiresAt           string                 `json:"expiresAt" gorm:"index"`
	LastResetAt         string                 `json:"lastResetAt"`
	NextResetAt         string                 `json:"nextResetAt" gorm:"index"`
	CreatedAt           string                 `json:"createdAt"`
	UpdatedAt           string                 `json:"updatedAt"`
}

type SubscriptionUsageLogType string

const (
	SubscriptionUsageConsume SubscriptionUsageLogType = "consume"
	SubscriptionUsageRefund  SubscriptionUsageLogType = "refund"
	SubscriptionUsageReset   SubscriptionUsageLogType = "reset"
)

type SubscriptionUsageLog struct {
	ID             string                   `json:"id" gorm:"primaryKey"`
	UserID         string                   `json:"userId" gorm:"index"`
	SubscriptionID string                   `json:"subscriptionId" gorm:"index"`
	Type           SubscriptionUsageLogType `json:"type" gorm:"index"`
	Amount         int                      `json:"amount"`
	Balance        int                      `json:"balance"`
	Model          string                   `json:"model"`
	Path           string                   `json:"path"`
	Remark         string                   `json:"remark"`
	CreatedAt      string                   `json:"createdAt"`
}

type CreditChargeStatus string

const (
	CreditChargePending  CreditChargeStatus = "pending"
	CreditChargeRefunded CreditChargeStatus = "refunded"
)

type CreditCharge struct {
	ID                  string             `json:"id" gorm:"primaryKey"`
	UserID              string             `json:"userId" gorm:"index"`
	SubscriptionID      string             `json:"subscriptionId" gorm:"index"`
	SubscriptionCredits int                `json:"subscriptionCredits"`
	WalletCredits       int                `json:"walletCredits"`
	TotalCredits        int                `json:"totalCredits"`
	Model               string             `json:"model"`
	Path                string             `json:"path"`
	Status              CreditChargeStatus `json:"status" gorm:"index"`
	CreatedAt           string             `json:"createdAt"`
	RefundedAt          string             `json:"refundedAt"`
}

type UserSubscriptionList struct {
	Items []UserSubscription `json:"items"`
	Total int                `json:"total"`
}

type SubscriptionSubscriber struct {
	SubscriptionID string                 `json:"subscriptionId"`
	UserID         string                 `json:"userId"`
	Username       string                 `json:"username"`
	DisplayName    string                 `json:"displayName"`
	QuotaCredits   int                    `json:"quotaCredits"`
	QuotaRemaining int                    `json:"quotaRemaining"`
	Status         UserSubscriptionStatus `json:"status"`
	StartsAt       string                 `json:"startsAt"`
	ExpiresAt      string                 `json:"expiresAt"`
}

type SubscriptionSubscriberList struct {
	Items []SubscriptionSubscriber `json:"items"`
	Total int                      `json:"total"`
}

type SubscriptionPurchaseResult struct {
	User         AuthUser         `json:"user"`
	Subscription UserSubscription `json:"subscription"`
}
